File: 0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include<unordered_map>
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head==NULL)
        {
            return false;
        }
       ListNode* slow=head;
        ListNode* fast=head->next;
        while(fast!=NULL&&fast->next!=NULL)
        {
            if(fast==slow)
            {
                return true;
            }
            slow=slow->next;
            fast=fast->next->next;
        }
        return false;
        
        // unordered_map<ListNode*,bool> visited;
        
        // ListNode* temp=head;
        // while(temp!=NULL)
        // { if(visited.find(temp)!=visited.end())
        // {
        //     return true;
        // }
        //  visited[temp]=true; 
        //  temp=temp->next;
        // }
        // return false;
        
    }
};

File: 0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
       if(head==null||head.next==null)
           return false;
        ListNode h=head;
        ListNode t=head;
        do{
            h=h.next.next;
            t=t.next;
            if(h==t)
            {
                return true;
            }
        }while(h!=null&&h.next!=null&&t!=null);
            return false;
    }
}